service/s3/s3manager/upload.go

```diff
@@ -373,6 +379,42 @@ func (u Uploader) UploadWithContext(ctx aws.Context, input *UploadInput, opts ..
 	return i.upload()
 }
 
+// UploadWithIterator will upload a batched amount of objects to S3. This operation uses
+// the iterator pattern to know which object to upload next. Since this is an interface this
+// allows for custom defined functionality.
+//
+// Example:
+//	svc:= s3manager.NewUploader(sess)
+//
+//	objects := []BatchUploadObject{
+//		{
+//			Object:	&s3manager.UploadInput {
+//				Key: aws.String("key"),
+//				Bucket: aws.String("bucket"),
+//			},
+//		},
+//	}
+//
+//	iter := &s3managee.UploadObjectsIterator{Objects: objects}
+//	if err := svc.UploadWithIterator(aws.BackgroundContext(), iter); err != nil {
+//		return err
+//	}
+func (u Uploader) UploadWithIterator(ctx aws.Context, iter BatchUploadIterator, opts ...func(*Uploader)) error {
+	objects := make([]BatchUploadObject, 0, u.BatchSize)
+
+	for iter.Next() {
+		objects = append(objects, iter.UploadObject())
+		if len(objects) >= u.BatchSize {
+			if err := uploadBatch(u, ctx, objects); err != nil {
+				return err
+			}
+			objects = objects[:0]
+		}
+	}
+
+	return uploadBatch(u, ctx, objects)
```

Comment:
It looks like these are trying to upload twice. What is the purpose of the BatchSize value since these uploads are all serial?  This is a little confusing design I'd imagine this could be simplified to something like the download with iterator method:


```go
func (u Uploader) UploadWithIterator(ctx aws.Context, iter BatchUploadIterator, opts ...func(*Uploader)) error {
    for iter.Next() {
        if _, err := svc.UploadWithContext(ctx, iter.UploadObject().Object); err != nil {
             return err
        }
   }
   return iter.Err()
}
```